### DISCUSSION TITLE
New keys at question level for Observation with rubrics implementation

### DISCUSSION THREAD

**Discussion by kiranharidas187** [Sunbird-inQuiry/Community#25] [General Discussions]
@vrayulu As part of the observation with rubrics flow we need to calculate the min and max score at each question and bubble it up to the criteria and domain level. 

For example, consider an MCQ question with 3 options of the different scores. Min score for the question will be the score of the option with the least score and the max will be the score of the option with the maximum score.

 

The score of criteria will be the sum of minimum scores of questions under the criteria and max score will be the sum of max scores of questions under the criteria. Similarly for the domain also the min score will be the sum of the minimum scores of criteria and the max score will be the sum of max scores of criteria under the domain.

 

In order to get this functionality done, we need a field known as minScore and maxScore in the question meta. Will this be fine?

cc- @vijiurs @joffinjoy @aks30 @pradeepkumarcm96

**Comment by vrayulu** [comment]
Will the minScore and maxScore for a question be auto-computed based on the scores assigned to each option or the creator has to mention it? and what will be the behaviour minScore and maxScore - if user gets less than the minScore, will there be any indication in the RESPONSE or ASSESS event or in the overall score of the question set?

@surendrasinghs @pallakartheekreddy - request your inputs also...

**Reply by kiranharidas187** [reply]
@vrayulu yes, the minScore and maxScore will be auto-calculated based on the scores entered at the option level. minScore and maxScore defines the minimum and maximum score the user will get on answering the question. For a non-mandatory question minScore will be zero and maxScore will depend on the option level score entered by the users.

This min and max score will be bubbled up to the criteria and further up till the domain. Then at criteria and domain, we define the levels.
<img width="386" alt="Screenshot 2022-08-05 at 2 21 17 PM" src="https://user-images.githubusercontent.com/21116904/183040834-58fe246f-620e-498b-a594-10f837d41650.png">

MinScore at criteria will be the sum of all min of  questions under it. Similarly for maxScore also.

Once a user submits the observation, each criteria and domain level will be calculated. For eg: So the net output will be like, for one particular **criteria** this particular school is at level **Good**

**Reply by vrayulu** [reply]
@kiranharidas187 - can you please think of other scenarios and define how minScore and maxScore will work in those scenarios? for example,
- when a question has partial scoring for options. e.g.: option 1 has 0.5 mark & option 2 has 0.5 mark..
- when a question is used in course assessments, will the assessment be marked as completed if user gets more than the overall min score? and what would user get if he/she scores the maxScore in the assessment (how will it be differentiated from just getting a minScore)?

if we want to make minScore & maxScore as first class elements of QuML, we need to have a generalised definition for those. May be, a discussion with inQuiry team (@surendrasinghs and @pallakartheekreddy ) will help in creating a generic definition for min & max score.

If that is not possible, please see if there are any other ways to implement this as an adopter/usecase specific configuration or extension for now.

**Comment by surendrasinghs** [comment]
Hi @kiranharidas187, 

Thanks for bringing this up. I believe this discussion belongs to the Sunbird QuML repository and not necessarily here at Sunbird inQuiry. Anyway, let’s think about what it means to add minScore and maxScore to the QuML Specification and enable it through Sunbird inQuiry.

MINSCORE and MAXSCORE are already available in the QuML specification. See https://github.com/sunbird-specs/QuML/blob/master/v1/question.md#question-variables 

What is the complete picture?: minScore and maxScore will work in conjunction with other pieces of the puzzle - taxonomy framework and rubric for assigning levels. Can you please help me understand the following:

1. How are you ensuring that questions are tagged with taxonomy (i.e. Criteria, Domain)?
2. How are you storing Criteria and Domain using the taxonomy framework?
3. How is the Rubric created for assigning levels? Is it created at the Question Set level? Please share any wireframes, specifications and implementation details.
4. How is the level computed and assigned (as per the rubric)? Is it on the device or on the server?
5. The attribute, say minScore and maxScore, would apply to all interaction types (i.e. single-choice, multi-choice, text, ordered, match, upload) and not just a particular interaction type (say, single-choice). So, imagine what would be minScore and maxScore behaviour for all these interactions?
6. So any future contributor trying to enable a new question interaction type also needs to build logic for minScore and maxScore computation during the creation and evaluation of a question. What support or guidance would you provide them (in the form of documentation, guidelines, or best practices)

@vrayulu @pallakartheekreddy Please add if anything.
